### ISSUE TITLE
o is null - OpenstackIdentityProvider / CloudFilesProvider

### ISSUE THREAD

**Issue by pedershk** [openstacknetsdk/openstack.net#550]
I am trying to use RunAbove blob storage (Swift/Openstack) with the SDK. I'm able to successfully authenticate against their API running Postman with the follow json: (returns 200 OK and a valid token)

```json
{
    "auth":{
        "tenantname":"redacted", 
        "passwordCredentials" : {
            "username":"henning@kilset.no", 
            "password":"redacted"
        }
    }
}
```

against https://auth.runabove.io/v2.0/tokens

I'm unable to do a normal "authenticate" with the OpenStackIdentityProvider though.

```c#

var authUrl = "https://auth.runabove.io/v2.0/";

var cloudIdentity = new CloudIdentityWithProject()
            {
                Username = "henning@kilset.no",
                Password = "redacted",
                ProjectName = "redacted"
            };

            var identityProvider = new OpenStackIdentityProvider(new Uri(authUrl), cloudIdentity);

            var authenticated = identityProvider.GetUserAccess(cloudIdentity);
```

Exception details:
```

System.ArgumentNullException was unhandled by user code
  HResult=-2147467261
  Message=Value cannot be null.
Parameter name: o
  ParamName=o
  Source=Newtonsoft.Json
  StackTrace:
       at Newtonsoft.Json.Linq.JToken.FromObjectInternal(Object o, JsonSerializer jsonSerializer)
       at net.openstack.Core.Providers.OpenStackIdentityProvider.<>c__DisplayClass60.<GetUserAccess>b__5f()
       at net.openstack.Core.Caching.UserAccessCache.<>c__DisplayClassc.<Get>b__4(String k)
       at System.Collections.Concurrent.ConcurrentDictionary`2.AddOrUpdate(TKey key, Func`2 addValueFactory, Func`3 updateValueFactory)
       at net.openstack.Core.Caching.UserAccessCache.Get(String key, Func`1 refreshCallback, Boolean forceCacheRefresh)
       at net.openstack.Core.Providers.OpenStackIdentityProvider.GetUserAccess(CloudIdentity identity, Boolean forceCacheRefresh)
       at Kilset.Redacted.Engine.Storage.SwiftMediaStorageProvider..ctor(IConfigurationHelper configuration, EngineDataContext dbContext) in c:\dev\redacted\src\Engine\Storage\SwiftMediaStorageProvider.cs:line 39
       at Kilset.Redacted.Integration.Test.Storage.SwiftMediaTests.Setup() in c:\dev\redacted\src\Integration.Test\Storage\SwiftMediaTests.cs:line 50
```

Not very helpful. I don't know what "o" refers to in this instance.

openstacknet 1.4.0.0

**Comment by carolynvs** [comment]
The error is caused by the code expecting both the ProjectName and ProjectId, even though the API really only needs one or the other... :stuck_out_tongue:  I will have a fix out soon and once it's released, will reply back here.

In the meantime the workaround is to specify both the ProjectName and the ProjectId, e.g.

```csharp
var cloudIdentity = new CloudIdentityWithProject()
            {
                Username = "{username}",
                Password = "{password}",
                ProjectName = "{project-name}",
                ProjectId = "{project-id}"
            };
```

**Comment by pedershk** [comment]
Thanks! That helped.

Now just to that annoying fact that I can't seem a find to set a region "per request". Instead having to set it on every call to the CloudFilesProvider. Should really be an (optional) constructor parameter. Something for a pull request, maybe?

**Comment by carolynvs** [comment]
A fix for this has been released in [v1.4.1](https://github.com/openstacknetsdk/openstack.net/releases/tag/v1.4.1). ProjectId is now optional.
